### DISCUSSION TITLE
Updating the pybricks hub firmware methods?

### DISCUSSION THREAD

**Discussion by johnscary-ev3** [pybricks/support#676] [Q&A]
I had to recently update the firmware on my Inventor hub to latest release.
It seems like the official way to do this is to first restore the LEGO firmware using restore method on install.pybricks.com and then use pybricksdev flash command to put new firmware on the hub.
This worked fine but is a bit indirect and so I am wondering about using the pybricksdev dfu command instead.
I got the dfu mode working, including adding the special driver, and made a backup file.
Seems like I could also use this mode to put the updated pybricks firmware onto the hub but was not sure if this was a good idea.
Is it ok to do the pybricks hub update this way with dfu command, and are there any things to watch out for with this?
Thanks.

By the way, pybricksdev flash command only seemed to work after I updated to Python 3.10 from Python 3.8.

**Comment by dlech** [comment]
> Is it ok to do the pybricks hub update this way with dfu command, and are there any things to watch out for with this?

Yes, although you will still use the `pybricksdev flash` command and not the `pybricksdev dfu` command. Just put the hub in DFU mode first (I'm assuming you already know how to do this). Then run `pybricksdev flash firmware.zip` and it will flash using DFU.

> By the way, pybricksdev flash command only seemed to work after I updated to Python 3.10 from Python 3.8.

This should be fixed in https://github.com/pybricks/pybricksdev/releases/tag/v1.0.0-alpha.27, if not please open an issue with the stack trace.

**Reply by johnscary-ev3** [reply]
Thanks for info on flash and dfu commands. I am still a bit confused on when to use each one.
Maybe some additional docs on this would be useful.
Also, it is not too clear when or if the LEGO firmware gets saved in this process or not using pybricksdev to update.
Not really a problem since we can get it back but just not sure how it is working. Thanks.

I am totally on Python 3.10 now and did not save the error trace I saw using pybricksdev a27 and Python 3.8.
However, I think it was the same as in the comment below.
I did not want to make new issue since I was not sure.

https://github.com/pybricks/support/issues/626#issuecomment-1150108025

**Reply by dlech** [reply]
We are still trying to figure out what is the best way for flashing firmware on these hubs. There are a number of technical issues that prevent if from being strait forward like hubs of the past.

The `pybricksdev dfu backup/restore` commands are strictly for making a backup of the firmware currently on the hub and restoring it later from the file created during the backup. The `pybricksdev flash` command flashes new firmware using a `firmware.zip` and can do things like customize the name of the hub.

**Reply by johnscary-ev3** [reply]
Thanks for clarification on when to use dfu vs flash commands.
Putting hub into dfu mode first and then using flash command is a bit confusing but I get it now. 

On that error I saw it is possible I still had pybricksdev a26 at that point.
I am still learning how to correctly update these various packages.

**Reply by dlech** [reply]
> I am still learning how to correctly update these various packages.

One thing I just learned is that you can use `pipx run pybricksdev ...` and it should always check for a new version pybricksdev and download it before running. This way you don't have to bother with manually checking for updates and installing the update yourself.
